lib/immobilienscout/response_parser.rb

```diff
@@ -33,13 +36,50 @@ def code
     end
 
     def messages(response_body)
-      return response_body unless response_body['common.messages']
+      if response_body['common.messages']
+        common_message(response_body)
+      elsif on_top_placement_type(response_body)
+        on_top_placement_message(response_body, on_top_placement_type(response_body))
+      else
+        response_body
+      end
+    end
 
+    def common_message(response_body)
       messages = response_body['common.messages'].first['message']
       messages = [messages] if messages.is_a? Hash
       messages.map { |msg| Message.new(nil, msg['messageCode'], msg['message'], msg['id']) }
     end
 
+    def on_top_placement_type(response_body)
+      if response_body['showcaseplacement.showcaseplacements']
+        :showcase_placement
+      elsif response_body['premiumplacement.premiumplacements']
+        :premium_placement
+      elsif response_body['topplacement.topplacements']
+        :top_placement
+      end
+    end
+
+    def on_top_placement_message(response_body, placement_type)
```

Comment:
For this method ChatGPT recommended the following changes to improve readability (I Agree with the recommendations 😆 )
```ruby
def on_top_placement_message(response_body, placement_type)
  is24_placement_type = Immobilienscout::API::OnTopPlacement.placement_type_for_is24(placement_type)

  placements = extract_placements(response_body, is24_placement_type)
  build_on_top_placement_messages(placements, placement_type)
end

def extract_placements(response_body, is24_placement_type)
  placements = response_body["#{is24_placement_type}.#{is24_placement_type.pluralize}"].first[is24_placement_type]
  placements.is_a?(Hash) ? [placements] : placements
end

def build_on_top_placement_messages(placements, placement_type)
  placements.map do |placement|
    OnTopPlacementMessage.new(
      placement['messageCode'],
      placement['message'],
      placement_type,
      service_period(placement, 'dateFrom'),
      service_period(placement, 'dateTo'),
      placement['@realestateid'],
      placement['externalId']
    )
  end
end

def service_period(placement, date_key)
  date_string = placement.dig('servicePeriod', date_key)
  date_string ? DateTime.parse(date_string) : nil
end
```